### PR TITLE
Changing gapi wrapper functions to include the completer logic

### LINKED ISSUE #81: Move the completer logic into javascript wrapper class

### PR DESCRIPTION
**Pull request by hanj7221**
Closes #81

### CODE REVIEW

File: test/root_component_test.dart [hunk 1]
```diff
@@ -87,15 +86,15 @@ void main() {
         'are invoked during init', () async {
       fixture = await testBed.create();
 
-      verify(mockCredential.handleClientLoad());
-      verify(mockCredential.initClient());
+      verify(mockGoogleApiDart.loadLibrary('client:auth2'));
+      verify(mockGoogleApiDart.initClient(any, any));
```
Comment (hanj7221) [inline test/root_component_test.dart]: I tried to verify that the correct input has been passed into the function by doing something like:

```
expect(verify(mockGoogleApiDart.initClient(captureAny, any)).captured, InitArgMatcher(initArgs));
```
but here is the error message i got:
```
  Expected: [object Object]
  Actual: NativeJavaScriptObject:<[object Object]>
```

Not sure if this is related to InitArgs having a `@JS()` annotator with it. Maybe we can take a look together when you have time.

**Review by hanj7221 — COMMENTED** [review]
This commit only changes the completer logic for functions related to credential validation. And I should do the same thing for the gapi query request functions, and I will make that a separate commit in this PR.

File: lib/src/credential_service.dart [hunk 2]
```diff
@@ -23,36 +22,30 @@ class CredentialService {
...
-    if (_googleAuth.isSignedIn.get()) {
-      _googleAuth
-          .signOut()
-          .then(allowInterop((_) => completer.complete(false)));
-      _googleAuth.disconnect();
-      return completer.future;
+  Future<bool> handleAuthClick() async {
+    if (googleAPIDart.getSignInStatus()) {
+      await googleAPIDart.signOut();
+      return Future.value(false);
```
Comment (ruben-aghayan) [inline lib/src/credential_service.dart]: do you have to wrap in a future here

File: lib/src/credential_service.dart [hunk 3]
```diff
@@ -23,36 +22,30 @@ class CredentialService {
   static const _dbmDiscoveryUrl =
       'https://content.googleapis.com/discovery/v1/apis/doubleclickbidmanager/v1.1/rest';
 
+  final GoogleApiDart googleAPIDart;
```
Comment (ruben-aghayan) [inline lib/src/credential_service.dart]: why public

File: test/root_component_test.dart (hunk 1 above)
Comment (ruben-aghayan) [reply test/root_component_test.dart]: sounds like the root cause. i think the arg to the googledartapi shouldnt be JS object. implementation should not be important to the caller. let googledartapi convert it into whatever form it needs to be for the js lib.

File: lib/src/credential_service.dart (hunk 2 above)
Comment (hanj7221) [reply lib/src/credential_service.dart]: Oops, looks like i don't

File: lib/src/credential_service.dart (hunk 3 above)
Comment (hanj7221) [reply lib/src/credential_service.dart]: sorry, should be private

File: lib/src/google_api_request_args.dart [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import 'package:built_value/built_value.dart';
+
+part 'google_api_request_args.g.dart';
+
+abstract class GoogleApiRequestArgs
+    implements Built<GoogleApiRequestArgs, GoogleApiRequestArgsBuilder> {
+  String get path;
+  String get method;
+  String get body;
```
Comment (hanj7221) [inline lib/src/google_api_request_args.dart:13]: The body field should have the  `@nullable` annotator with it.
Tests update are still pending and will be submitted in a separate commit in this PR.

**Review by ruben-aghayan — APPROVED** [review]

File: test/query_component_test.dart [hunk 5]
```diff
@@ -578,3 +626,183 @@ void main() {
     });
   });
 }
+
+class isDV3RequestWith extends Matcher {
```
Comment (ruben-aghayan) [inline test/query_component_test.dart:569]: nice!

File: test/query_component_test.dart [hunk 6]
```diff
@@ -253,17 +241,16 @@ void main() {
...
-            .thenAnswer((_) => reportCompleter.future);
+        final answers = [
+          Future.value(multipleIOJsonString),
+          Future.value(reportingCreateQueryApiJsonResponse),
+          Future.value(reportingGetQueryApiJsonResponse),
+          reportCompleter.future,
+        ];
+
+        var callCount = 0;
+        when(mockGoogleApiDart.request(any))
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: please list them out explicitly
e.g. when(mockApi.request(specificRequest).thenAnswer((_) => specificAnswer)

File: test/query_component_test.dart [hunk 7]
```diff
@@ -277,17 +264,17 @@ void main() {
...
       });
 
       test('non-integer advertiser id disables the populate button', () async {
         await queryComponentAccordionPO.typeAdvertiserId('wrong-format-id');
         await queryComponentPO.clickPopulate();
-        verifyNever(mockQueryService.execDV3Query(any, any, any, any, any));
+        verifyNever(mockGoogleApiDart.request(any));
       });
 
       test('highlighting unchecked invokes populate(multiple-IO, false)',
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: are you testing the checkbox or all of the queries? I suggest you confirm the queries once in one test, and in 2 other test confirm the checkbox behavior. this way there are fewer duplicated asserts and it is more clear what the test is testing.

File: test/query_component_test.dart [hunk 8]
```diff
@@ -353,17 +360,16 @@ void main() {
...
-            .thenAnswer((_) => Future.value(multipleIOJsonString));
-        when(mockQueryService.execReportingCreateQuery(QueryType.byMediaPlan,
-                advertiserId, mediaPlanId1, argThat(isNull), startDate, any))
-            .thenAnswer(
-                (_) => Future.value(reportingCreateQueryApiJsonResponse));
-        when(mockQueryService.execReportingGetQuery(queryId))
-            .thenAnswer((_) => Future.value(reportingGetQueryApiJsonResponse));
-        when(mockQueryService.execReportingDownload(downloadLink))
-            .thenAnswer((_) => reportCompleter.future);
+        final answers = [
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: same

File: test/query_component_test.dart [hunk 9]
```diff
@@ -470,61 +500,55 @@ void main() {
...
-                ioId1,
-                startDate,
-                any))
-            .thenAnswer(
-                (_) => Future.value(reportingCreateQueryApiJsonResponse));
-        when(mockQueryService.execReportingGetQuery(queryId))
-            .thenAnswer((_) => Future.value(reportingGetQueryApiJsonResponse));
-        when(mockQueryService.execReportingDownload(downloadLink))
-            .thenAnswer((_) => reportCompleter.future);
+        final answers = [
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: same throughout

File: test/query_component_test.dart [hunk 10]
```diff
@@ -470,61 +500,55 @@ void main() {
...
 
       test('non-integer io id disables the populate button', () async {
         await queryComponentAccordionPO.typeAdvertiserId(advertiserId);
         await queryComponentAccordionPO.typeInsertionOrderId('wrong-format-id');
         await queryComponentPO.clickPopulate();
-        verifyNever(mockQueryService.execDV3Query(any, any, any, any, any));
+        verifyNever(mockGoogleApiDart.request(any));
       });
 
       test('highlighting unchecked invokes populate(single-IO, false)',
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: no need to test every combination of query type + checkbox
1 test per query type, and 1 test for each checkbox state (i.e. 2)

**Review by ruben-aghayan — CHANGES_REQUESTED** [review]

File: test/query_component_test.dart [hunk 11]
```diff
@@ -224,6 +224,58 @@ void main() {
...
+
+      tearDown(() => clearInteractions(mockGoogleApiDart));
+
+      test('is selected, populate(any, true) is invoked', () async {
+        await queryComponentAccordionPO.typeAdvertiserId(advertiserId);
+        await queryComponentAccordionPO.typeInsertionOrderId(ioId1);
+        await queryComponentPO.selectUnderpacing();
+        await queryComponentPO.clickPopulate();
+
+        // waits for the last operation execReportingDownload() to finish.
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: rm comment

File: test/query_component_test.dart [hunk 12]
```diff
@@ -277,47 +333,15 @@ void main() {
...
-                QueryType.byAdvertiser, advertiserId))));
-        verify(mockGoogleApiDart
-            .request(argThat(isReportingGetQueryRequestWith(queryId))));
-        verify(mockGoogleApiDart
-            .request(argThat(isReportingDownloadRequestWith(downloadLink))));
-        verify(mockExcelDart.populate(expectedInput, false));
-      });
-
-      test('highlighting checked invokes populate(multiple-IO, true)',
+      test('highlighting unchecked invokes populate() with correct list of IOs',
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: update name (ie checkbox is irrelevant)

File: test/query_component_test.dart [hunk 13]
```diff
@@ -277,47 +333,15 @@ void main() {
...
-            argThat(isDV3RequestWith(QueryType.byAdvertiser, advertiserId))));
-        verify(mockGoogleApiDart.request(argThat(
-            isReportingCreateQueryRequestWith(
-                QueryType.byAdvertiser, advertiserId))));
-        verify(mockGoogleApiDart
-            .request(argThat(isReportingGetQueryRequestWith(queryId))));
-        verify(mockGoogleApiDart
-            .request(argThat(isReportingDownloadRequestWith(downloadLink))));
-        verify(mockExcelDart.populate(expectedInput, true));
+        verify(mockExcelDart.populate(expectedInput, any));
```
Comment (ruben-aghayan) [inline test/query_component_test.dart]: construct expectation inside test (i dont think its reused?)

**Review by ruben-aghayan — APPROVED** [review]
